packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp

```diff
@@ -572,109 +577,172 @@ void CoalesceDropFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level
               }
               rows(row + 1) = realnnz;
             }
-          } else {
-            /* Cut Algorithm */
-            // CMS
-            using DropTol = Details::DropTol<real_type, LO>;
-            std::vector<DropTol> drop_vec;
-            drop_vec.reserve(nnz);
-            const real_type zero = Teuchos::ScalarTraits<real_type>::zero();
-            const real_type one  = Teuchos::ScalarTraits<real_type>::one();
-            LO rownnz            = 0;
-            // NOTE: This probably needs to be fixed for rowsum
-
-            // find magnitudes
-            for (LO colID = 0; colID < (LO)nnz; colID++) {
-              LO col = indices[colID];
-              if (row == col) {
-                drop_vec.emplace_back(zero, one, colID, false);
-                continue;
+          }  // end for row
+        }
+        else {
+          /* Cut Algorithm */
+          SubFactoryMonitor m1(*this, "Cut Drop", currentLevel);
+          using ExecSpace = typename Node::execution_space;
+          using TeamPol = Kokkos::TeamPolicy<ExecSpace>;
+          using TeamMem = typename TeamPol::member_type;
+          using ATS = Kokkos::ArithTraits<Scalar>;
+          using impl_scalar_type = typename ATS::val_type;
+          using implATS = Kokkos::ArithTraits<impl_scalar_type>;
+
+          //move from host to device
+          auto ghostedDiagValsView = Kokkos::subview(ghostedDiag->getDeviceLocalView(Xpetra::Access::ReadOnly), Kokkos::ALL(), 0);
+          auto boundaryNodesDevice = Kokkos::create_mirror_view(ExecSpace(), boundaryNodes);
+          auto thresholdKokkos = static_cast<impl_scalar_type>(threshold);
+          auto realThresholdKokkos = implATS::magnitude(thresholdKokkos);
+
+          auto At = Utilities::Op2TpetraCrs(A);
+          auto A_device = At->getLocalMatrixDevice();
+
+          int algorithm = classicalAlgo;
+          Kokkos::View<LO*, ExecSpace>rownnzView("rownnzView", A_device.numRows());
+          auto drop_views = Kokkos::View<bool*, ExecSpace>("drop_views", A_device.nnz());
+          auto index_views = Kokkos::View<size_t*, ExecSpace>("index_views", A_device.nnz());
+
+          Kokkos::parallel_reduce("classical_cut", TeamPol(A_device.numRows(), Kokkos::AUTO), KOKKOS_LAMBDA(const TeamMem& teamMember, LO& globalnnz, GO& totalDropped) {
+            LO row = teamMember.league_rank();
+            auto rowView = A_device.row(row);
+            size_t nnz = rowView.length;
+
+            size_t n = 0;
```

Comment:
Rename to `dropStart`?

Reply:
`n` is meant to represent the number of entries in `drop_view` while `dropStart` is meant to represent the index of where dropping should start. Thus, the two variable should be kept separate. Would renaming `n` to `dropSize` make this more clear?

Reply:
Oh, I see now that `n` is used again further down. Yes, let's rename it to `dropSize` to improve readability.